core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -48,32 +56,69 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
   // of subDirs(i) is protected by the lock of subDirs(i)
   private val subDirs = Array.fill(localDirs.length)(new Array[File](subDirsPerLocalDir))
 
+  private[spark] val badDirs = ArrayBuffer[File]()
+  private[spark] val dirToBlacklistExpiryTime = new HashMap[File, Long]
+  // Filename hash to dirId, it should be small enough to put into memory
+  private[spark] val migratedDirIdIndex = new ConcurrentHashMap[Int, Int].asScala
+
   private val shutdownHook = addShutdownHook()
 
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
   def getFile(filename: String): File = {
+    var mostRecentFailure: Exception = null
     // Figure out which local directory it hashes to, and which subdirectory in that
     val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
+    val dirId = migratedDirIdIndex.getOrElse(hash, hash % localDirs.length)
```

Comment:
I think we cannot use hash code as a key for the `migratedDirIdIndex` map (as if a filename is already mapped then always the same FILE instance should be given back for it).

Example: Let's assume `alpha` and `beta` are two valid filenames with the the same hash code.
Let's say first we call `getFile(alpha)` and there is no corruption (a FILE is given back and  some content is written for storing a block) then we call `getFile(beta)` where corruption is detected so `badDirs` and `migratedDirIdIndex` are updated. Now comes a reading from `getFile(alpha)` which  because of the `migratedDirIdIndex` gives back a new FILE instance with empty content.

You can extend your test with some reading and writing of contents.

Reply:
Collisions can happen, like
```
scala> val s = for (i <- 1 to 100; j <- 1 to 100; k <- 1 to 100) yield Math.abs(("shuffle_" + i + "_" + j + "_" + k + ".index").hashCode)
scala> s.size
res29: Int = 1000000

scala> s.toSet.size
res30: Int = 999964
```